enhancements/kube-apiserver/audit-policy.md

```diff
@@ -261,61 +279,294 @@ the respective file-based audit policy of
+type AuditCustomRule struct {
+    // group is a name of group a request user must be member of in order to this profile to apply.
+    //
+    // +kubebuilder:validation:Required
+    // +kubebuilder:validation:MinLength=1
+    // +required
+    Group string `json:"group"`
+
+    // profile specifies the name of the desired audit policy configuration to be deployed to
+    // all OpenShift-provided API servers in the cluster.
+    //
+    // The following profiles are provided:
+    // - Default: the existing default policy.
+    // - WriteRequestBodies: like 'Default', but logs request and response HTTP payloads for
+    // write requests (create, update, patch).
+    // - AllRequestBodies: like 'WriteRequestBodies', but also logs request and response
+    // HTTP payloads for read requests (get, list).
+    // - None: no requests are logged at all, not even oauthaccesstokens and oauthauthorizetokens.
+    //
+    // If unset, the 'Default' profile is used as the default.
+    //
+    // +kubebuilder:Enum=Default,WriteRequestBodies,AllRequestBodies,None
+    // +kubebuilder:validation:Required
+    // +required
+    Profile AuditProfileType `json:"profile,omitempty"`
+}
+```
+
+### Library-go Implementation of CustomRules
+
+The library-go interface for audit policies before addition of `customRules` in `github.com/openshift/library-go/pkg/operator/apiserver/audit` looked like this:
+
+```go
+type AuditPolicyPathGetterFunc func(profile string) (string, error)
+
+func DefaultPolicy() ([]byte, error)
+func WithAuditPolicies(targetName string, targetNamespace string, assetDelegateFunc resourceapply.AssetFunc) resourceapply.AssetFunc
+func GetAuditPolicies(targetName, targetNamespace string) (*corev1.ConfigMap, error)
+func NewAuditPolicyPathGetter(path string) (AuditPolicyPathGetterFunc, error)
+```
+
+with a config observer in `github.com/openshift/library-go/pkg/operator/configobserver/apiserver`:
+
+```go
+func NewAuditObserver(pathGetter AuditPolicyPathGetterFunc) configobserver.ObserveConfigFunc
+```
+
+The asset func is wired into the static resource controller in our operators like this:
+
+```go
+apiservercontrollerset.NewAPIServerControllerSet(
+    ...
+).WithStaticResourcesController(
+    "APIServerStaticResources",
+    libgoassets.WithAuditPolicies("audit", operatorclient.TargetNamespace, v311_00_assets.Asset),
+    []string{
+	    ...,
+		libgoassets.AuditPoliciesConfigMapFileName,
+	},
+	...,
+).WithRevisionController(
+	operatorclient.TargetNamespace,
+	[]revision.RevisionResource{
+		{
+			Name: "audit",
+		},
+	},
+	...,
+)
+```
+
+As audit policy file is not static anymore with the addition of custom rules, this approach will not work anymore. Instead, we need
+a dynamic audit policy controller that computes the policy depending on the apiserver configuration resource and copies that as a
+`ConfigMap` into the operand namespace. Then the revision controller will notice it changing and assigns a new operand revision that
```

Comment:
I could be wrong, but I thought the file was for the operand/apiserver and the ConfigMap was for the operators.

Reply:
the configmap is created by the operator in the operand namespace. The revision controller makes a copy with the `-<revision>` suffix and redeploys the apiserver with the new configmap mounted.

Reply:
That clears it up. So it's operator initiated and consumed by operand 👍

Reply:
yep.